### PR TITLE
Adapt to Starknet / cairo-lang 0.12.0

### LINKED ISSUE #511: Support Starknet / cairo-lang 0.12
At the time of writing this, there is not pre-release at https://pypi.org/project/cairo-lang/#history

Implies supporting Cairo 2 (rc4?)

### LINKED ISSUE #514: In blocks-on-demand mode, all transactions have transaction_index 0
They shouldn't. First transaction in a block should have index, second should have index 1, etc.

Reproduce by:
1. `starknet-devnet --blocks-on-demand`
2. Send a couple of transactions
3. Retrieve the pending block
    - `curl localhost:5050/feeder_gateway/get_block?blockNumber=pending` (your URL may vary)

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Close #511
- Close #514
- Replace status code 500 with 400
  - cairo-lang introduced this and due to the way we handle some errors, the code would be changed only for those instances where we 
- Blocks can still be pending, but the transactions in it are marked as `ACCEPTED_ON_L2`

## Development related changes

- Rewrite testing contracts to cairo 2
- Stop generating ABI for Cairo 1 contracts in tests - not needed for interaction
- According to cairo-lang requirements, add `StorageDomain` to nonce and storage manipulation (some state methods now require the new `storage_domain` param)
- Forking tests are not passing - expected because testnet (used in forking tests) hasn't been updated (only integration has been updated)

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [ ] All tests are passing - `./scripts/test.sh`

### CODE REVIEW

File: starknet_devnet/devnet_config.py [hunk 1]
```diff
@@ -50,9 +50,9 @@
 DEFAULT_CHAIN_ID = StarknetChainId.TESTNET
 
 DEFAULT_COMPILER_ARGS = [
+    # without the space at the beginning, our tests fail because it gets
```
Comment (mikiw) [inline starknet_devnet/devnet_config.py]: 😮

File: starknet_devnet/account_util.py [hunk 2]
```diff
@@ -120,12 +121,18 @@ async def set_balance(state: StarknetState, address: int, balance: int):
...
+        storage_domain=StorageDomain.ON_CHAIN,
+        contract_address=fee_token_address,
+        key=balance_key,
+        value=balance_uint256.low,
     )
     await state.state.set_storage_at(
-        fee_token_address, balance_address + 1, balance_uint256.high
+        storage_domain=StorageDomain.ON_CHAIN,
+        contract_address=fee_token_address,
+        key=balance_key + 1,
```
Comment (mikiw) [inline starknet_devnet/account_util.py:136]: I don't understand this `balance_key + 1`. What is that?

File: starknet_devnet/fee_token.py [hunk 3]
```diff
@@ -58,24 +59,28 @@ def contract_class(self) -> DeprecatedCompiledClass:
     async def _mimic_constructor(self):
         starknet: Starknet = self.starknet_wrapper.starknet
         await starknet.state.state.set_storage_at(
-            FeeToken.ADDRESS,
-            get_selector_from_name("ERC20_name"),
-            str_to_felt(FeeToken.NAME),
+            storage_domain=StorageDomain.ON_CHAIN,
```
Comment (mikiw) [inline starknet_devnet/fee_token.py:62]: StorageDomain.ON_CHAIN means data on L1? Shouldn't we use StorageDomain.OFF_CHAIN? Shouldn't we also mention that `StorageDomain.ON_CHAIN` in the documentation?

Is that even important? Maybe not, I'm just asking.

**Review by mikiw — APPROVED** [review]
besides my questions/concerns PR looks fine

File: starknet_devnet/fee_token.py (hunk 3 above)
Comment (FabijanC) [reply starknet_devnet/fee_token.py:62]: In cairo-lang, there is an internal assertion on the used storage domain - they assert that it's on_chain.

https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/business_logic/state/storage_domain.py

File: starknet_devnet/account_util.py (hunk 2 above)
Comment (FabijanC) [reply starknet_devnet/account_util.py:138]: This was not added in this PR, but explained in a comment: https://github.com/0xSpaceShard/starknet-devnet/pull/512/commits/5094cfd6c732ae6d683361b46e1059520894dd08

**Pull request merged**